### ISSUE TITLE
cxxtools 3.0: failed to compile on Fedora 34 with gcc11

### ISSUE THREAD

**Issue by martinkg** [maekitalo/cxxtools#27]
Hi,

i am trying to compile cxxtools 3.0 on Fedora 34 with gcc11 but this fails with following error messages:

```
In file included from /usr/include/c++/11/bits/basic_string.h:48,
                 from /usr/include/c++/11/string:55,
                 from ../include/cxxtools/char.h:32,
                 from settingswriter.h:31,
                 from settingswriter.cpp:28:
/usr/include/c++/11/string_view: In instantiation of 'class std::basic_string_view<cxxtools::Char, std::char_traits<cxxtools::Char> >':
settingswriter.cpp:42:26:   required from here
/usr/include/c++/11/string_view:98:21: error: static assertion failed
   98 |       static_assert(is_trivial_v<_CharT> && is_standard_layout_v<_CharT>);
      |                     ^~~~~~~~~~~~~~~~~~~~
/usr/include/c++/11/string_view:98:21: note: 'std::is_trivial_v<cxxtools::Char>' evaluates to false
make[2]: *** [Makefile:920: settingswriter.lo] Error 1
```
[Build.log](https://kojipkgs.fedoraproject.org//work/tasks/4046/60084046/build.log) on Fedora build server

**Comment by jwakely** [comment]
`cxxtools::Char` has a non-trivial default constructor, which means it's not a trivial type.

The C++ standard says that `std::string` and `std::string_view` are "components for manipulating sequences of any non-array trivial standard-layout type." It is undefined to use them to store non-trivial types, which is why I added the static assertion to GCC's standard library.

**Comment by jwakely** [comment]
> i am trying to compile cxxtools 3.0 on Fedora 34 with c++11 but this fails with following error messages:

@martinkg do you mean GCC 11 not c++11? Because you're compiling it as C++17 not C++11 (otherwise `std::string_view` wouldn't be usable).

**Comment by martinkg** [comment]
sorry, it's gcc-11.0.0

**Comment by maekitalo** [comment]
Uh, that is really a major issue. It needs major changes. Very bad. I need to look at it, but it will take some time.

**Comment by jwakely** [comment]
Defining the `Char()` default ctor as defaulted will allow it to be used in `string_view`, but of course you need to check everywhere that uses default init of a `Char` and replace it with value-init.

**Comment by maekitalo** [comment]
A I see. I don't quite get it, why I need to specify the Char::Char() as default, but it looks like it works that way. I just do not have a gcc 11 by hand but at least std::is_trival<char::Char>::value returns true then. I will check that in.

**Comment by jwakely** [comment]
A class with a user-provided default constructor is not a trivial class.

By defining the default constructor as defaulted the compiler defines it, and it is a trivial constructor (it doesn't initialize anything) and because the copy constructor and copy assignment operator and destructor are already trivial, that allows the class to be a trivial class.

**Comment by martinkg** [comment]
if you want, provide a patch that will allow me to compile cxxtools with the gcc11.

**Comment by maekitalo** [comment]
I fixed it. Can you take the commit b773c01fc13d2ae67abc0839888e383be23562fd as a patch? Very trivial.

**Comment by martinkg** [comment]
the issue was fixed with commit b773c01.
Many Thanks
